README
********************************************
Readme / documentation for the execo package
********************************************

Execo offers a Python API for asynchronous control of local or remote,
standalone or parallel, unix processes. It is especially well suited
for quickly and easily scripting workflows of parallel/distributed
operations on local or remote hosts: automate a scientific workflow,
conduct computer science experiments, perform automated tests,
etc. The core python package is ``execo``. The ``execo_g5k`` package
provides a set of tools and extensions for the `Grid5000
<https://www.grid5000.fr>`_ testbed. The ``execo_engine`` package
provides tools to ease the development of computer sciences
experiments.

License
=======

Execo is copyright INRIA Rhone-Alpes, Service Experimentation et
Developpement.

Execo is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your
option) any later version.

Execo is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with Execo.  If not, see <http://www.gnu.org/licenses/>

Versions
========

latest stable version: v2.6.3 (2017-11-29)

Installation instructions
=========================

useful links
------------

- home page:  http://execo.gforge.inria.fr

- documentation pages:

  - latest stable release:
    http://execo.gforge.inria.fr/doc/latest-stable

  - development version: http://execo.gforge.inria.fr/doc/dev

- package download: http://execo.gforge.inria.fr/downloads or
  https://gforge.inria.fr/frs/?group_id=2901

- mailing lists:

  - execo-users@lists.gforge.inria.fr for execo users
    (https://lists.gforge.inria.fr/mailman/listinfo/execo-users)

  - execo-devel@lists.gforge.inria.fr to discuss execo development
    (https://lists.gforge.inria.fr/mailman/listinfo/execo-devel)

  - execo-commits@lists.gforge.inria.fr to be notified of commits
    (https://lists.gforge.inria.fr/mailman/listinfo/execo-commits)

- bug tracker: http://bugzilla.inria.fr - product: execo

- code repository: git://scm.gforge.inria.fr/execo/execo.git

- project page on INRIA Forge: http://gforge.inria.fr/projects/execo/

platforms
---------

- works on linux (primary development platform), darwin (macosx)

- should work on bsd (not tested)

- don't know on windows or cygwin (not tested)

prerequisites
-------------

execo installation absolutely requires ``python`` 2.6 / 2.7 or
>=3.2. ``execo_g5k`` needs ``python-requests``. Optionnal packages are
(debian package names, in order of decreasing importance):

- under python 2.6, ``python-argparse`` is needed for
  `execo_engine.engine.Engine`

- ``python-keyring`` for allowing storage of `execo_g5k.api_utils` api
  password in the desktop environment keyring (allowing asking it only
  once).

- ``python-psycopg2`` for optimized interface to Grid5000 OAR planning.

- ``sphinx-doc``, ``graphviz`` for building the documentation (usually
  not needed for regular users).

- ``python-matplotlib`` (>= 1.2.0) for some graphical representations.

- ``python-networkx``, ``python-pygraphviz``, ``graphviz`` for module
  ``execo_g5k.topology``. Later versions of networkx (at least v1.11)
  also require ``python-pydotplus``, but this dependency is not
  explicitely listed in the debian package (as of July 2017).

At runtime, connecting to remote hosts requires ``ssh``, ``scp`` or
similar connection tools, and optionnaly ``taktuk`` (probably >=
3.6. http://taktuk.gforge.inria.fr/). `execo.action.ChainPut` requires
bourne shell and ``netcat`` on remote hosts.

installation
------------

To install execo from a source package (.tar.gz) or from the git
tree::

 $ python setup.py install [--user]

to build documentation (if sphinx is available)::

 $ python setup.py build_doc

to install documentation (if sphinx is available)::

 $ python setup.py install_doc

It is possible to install execo automatically with ``pip`` or
``easy_install``::

 $ pip install --user execo

or::

 $ easy_install --user execo

It is possible to generate a debian package. For example, in the execo
package toplevel dir, run (the .deb will be generated in ../)::

 $ debuild --no-tgz-check -b -i -I -k<signing key id>

Usage
=====

See html documentation for module execo and execo_g5k at
http://execo.gforge.inria.fr/doc/

Bugs
====

Execo is regularly used to perform advanced experiments and
administration / monitoring tasks, in and outside Grid5000. We
actively fix bugs. Bugs should be reported to http://bugzilla.inria.fr
(product: execo)

Publications
============

Matthieu Imbert, Laurent Pouilloux, Jonathan Rouzaud-Cornabas, Adrien
Lèbre, Takahiro Hirofuchi "`Using the EXECO toolbox to perform
automatic and reproducible cloud experiments
<http://hal.inria.fr/hal-00861886/>`_" *1st International Workshop on
UsiNg and building ClOud Testbeds UNICO, collocated with IEEE CloudCom
2013* 2013

How to contribute
=================

- start contributing by sending clean patches or report bugs.

- stay consistent with the coding and naming style

- code must be compatible with python 2.6, 2.7 and 3.2+ (see
  http://execo.gforge.inria.fr/doc/latest-stable/python2-python3.html)

- use the core systems provided (eg. the configuration system, the tty
  coloring system, the logger, etc.) instead of developing your own.

- provide documented code (internal documentation as well as user
  documentation where needed)

- ask a core developer before adding a dependency or dealing with
  threads, signals (particularly: creating threads)

- indent with spaces, not tabs. one level of indentation is four
  spaces (if needed: use ``reindent.py -rnv .`` in execo top
  directory)

- recommended commit messages format:

  ``[<module_name>] <category>: commit message``.

  - <module_name> can be execo, execo_g5k, execo_engine. Omit if
    commit is not specific to a module.

  - <category> is free form but should indicate which part(s) of the
    module the commit affects. Omit if a commit is not specific to a
    module part.


debian/changelog
python-execo (2.6.3) unstable; urgency=low

  * new upstream release:

    - add vagrant vm + scripts for building debian wheezy compatible
    packages

    - major fixes and improvements in topology

    - fix various python 3 compatibility issues

    - engine: fix doc

    - various other fixes and improvements

 -- Matthieu Imbert <matthieu.imbert@inria.fr>  Wed, 29 Nov 2017 16:48:00 +0100

python-execo (2.6.2) unstable; urgency=low

  * new upstream release:

    - fix process output decoding for python3 <= 3.4

    - ProcessOutputHandler: fix line handling issue

    - switch from httplib2 to requests

    - fix optparse to argparse migration issue

    - remove unneeded warning in topology

    - fix and improve get_hosts_metric

 -- Matthieu Imbert <matthieu.imbert@inria.fr>  Fri, 9 Jun 2017 10:30:00 +0200

python-execo (2.6.1) unstable; urgency=low

  * new upstream release:

    - fix debian packaging for python2 / python3

    - make it packageable under wheezy

 -- Matthieu Imbert <matthieu.imbert@inria.fr>  Tue, 11 Oct 2016 16:22:00 +0200

python-execo (2.6) unstable; urgency=low

  * new upstream release:

    - Compatible with python 2.6, 2.7, >3.2. A few API have changed due to
    this major improvement.

    - Major optimization of Process output handling when dealing with
    outputs with lots of lines

    - execo_g5k.topology: fix lack of discrimination of links between
    different interfaces of a node and different linecards of a single
    network equip

    - execo_g5k: use nb_cores instead of smt_size to accommodate the
    change in g5k api

    - execo_g5k.api_utils: add get_hosts_metric() and remove
    get_g5k_measures()

    - execo_g5k.api_utils: get_oar_job_kavlan() now returns a list

    - execo_g5k api_utils add set_nodes_vlan

    - misc bugfixes and improvements

 -- Matthieu Imbert <matthieu.imbert@inria.fr>  Wed, 5 Oct 2016 15:25:00 +0200

python-execo (2.5.4) unstable; urgency=low

  * new upstream release:

    - execo_g5k.topology: fix wrong log template causing crash, log
    messages, documentation

    - execo_g5k.planning: handle oar queues in planning

    - execo_g5k.charter: fix bad sql results indexing causing crash

    - make execo / execo_g5k runnable with void environment (no $HOME)

    - execo.process: important optimization: bypass heavy string
    processing of iodebug logger when possible

 -- Matthieu Imbert <matthieu.imbert@inria.fr>  Tue, 3 May 2016 17:00:00 +0200

python-execo (2.5.3) unstable; urgency=low

  * new upstream release:

    - execo_g5k.topology: fix crash with empty linecards

    - execo_g5k.oarsub: fix oar command in corner cases

 -- Matthieu Imbert <matthieu.imbert@inria.fr>  Mon, 25 Jan 2016 15:31:00 +0200

python-execo (2.5.2) unstable; urgency=low

  * new upstream release:

    - bugfixes in timeouts handling

    - bugfixes and improvements in the auto force killing mechanism

    - planning: resources in maintenance mode

    - exact same behavior of api planning or postgresql planning

    - planning fix corner cases bugs with empty sites (no resources)

    - planning fix sites not responding handled correctly

    - planning add besteffort option

    - add filtering of g5k clusters on oar queues (allowing auto selection
    of the relevant clusters by default)

    - bugfixes in port forwarders (timeout handling)

 -- Matthieu Imbert <matthieu.imbert@inria.fr>  Tue, 1 Dec 2015 09:34:00 +0200

python-execo (2.5.1) unstable; urgency=low

  * new upstream release:

    - update execo_g5k to postgresql (due to g5k migration from mysql to
    postgres)

    - execo_engine.ParamSweeper: add batched versions of combinations API

 -- Matthieu Imbert <matthieu.imbert@inria.fr>  Wed, 30 Sep 2015 13:29:00 +0200

python-execo (2.5) unstable; urgency=low

  * new upstream release:

    - User-friendly way to write to processes stdin. Processes and Remotes
    can be used as file-like objects. Additionally, process stdin is not
    anymore closed by default.

    - Automatic ignoring of process errors when explicitly killing them

    - Add processes and actions context managers

    - Add special processes and actions for communicating with local or
    remote (through ssh) serial ports

    - Add expect-like syntax to processes and actions (for easy scripting
    of interactive local/remote/serial sessions)

    - The return of the revenge of the infamous keyword arguments

    - On-disk caching of most of the g5k api, for fast api requests

    - Add multigraph support to g5k topology

    - Add handling of g5k network equipments in api_utils

    - Internal taktuk interface bugfixes

 -- Matthieu Imbert <matthieu.imbert@inria.fr>  Wed, 02 Sep 2015 10:10:00 +0200

python-execo (2.4.3) unstable; urgency=low

  * new upstream bugfix release: again fix pip packaging

 -- Matthieu Imbert <matthieu.imbert@inria.fr>  Thu, 21 Nov 2014 08:48:00 +0100

python-execo (2.4.2) unstable; urgency=low

  * new upstream bugfix release: fix pip packaging

 -- Matthieu Imbert <matthieu.imbert@inria.fr>  Thu, 20 Nov 2014 14:50:00 +0100

python-execo (2.4) unstable; urgency=low

  * Initial debian packaging release

 -- Matthieu Imbert <matthieu.imbert@inria.fr>  Thu, 20 Nov 2014 10:46:00 +0100


src/execo/config.py
# Copyright 2009-2017 INRIA Rhone-Alpes, Service Experimentation et
# Developpement
#
# This file is part of Execo.
#
# Execo is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Execo is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
# License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Execo.  If not, see <http://www.gnu.org/licenses/>

from __future__ import print_function
import logging
import os
import sys

SSH = 0
SCP = 1
TAKTUK = 2
CHAINPUT = 3

def checktty(f):
    try:
        if (get_ipython().__class__.__name__ == 'ZMQInteractiveShell' #@UndefinedVariable
            and f.__class__.__module__ == 'IPython.zmq.iostream'
            and f.__class__.__name__ == 'OutStream'):
            return True
    except NameError:
        pass
    if hasattr(f, 'isatty'):
        return f.isatty()
    if hasattr(f, 'fileno'):
        return os.isatty(f.fileno())
    return False

IODEBUG = 3
FDEBUG = 5
TRACE = 12
DETAIL = 15

# _STARTOF_ configuration
configuration = {
    'log_level': logging.INFO,
    'remote_tool': SSH,
    'fileput_tool': SCP,
    'fileget_tool': SCP,
    'compact_output_threshold': 4096,
    'kill_timeout': 60,
    'intr_period': 1,
    'port_range': (25500, 26700),
    'kill_childs_at_end': True,
    'color_mode': checktty(sys.stdout)
                  and checktty(sys.stderr),
    'color_styles': {
        'log_header': ('yellow',),
        'object_repr': ('blue',),
        'emph': ('cyan',),
        'report_warn': ('magenta',),
        'report_error': ('red', 'bold'),
        'command': ('blue', 'bold'),
        'host': ('magenta', 'bold'),
        'user1': ('green', 'bold'),
        'user2': ('yellow', 'bold'),
        'user3': ('cyan', 'bold'),
        IODEBUG: ('green', 'reverse'),
        FDEBUG: ('green', 'reverse'),
        logging.DEBUG: ('green',),
        TRACE: ('green', 'bold',),
        DETAIL: ('magenta', 'bold'),
        logging.INFO: ('magenta',),
        logging.WARNING: ('cyan',),
        logging.ERROR: ('red',),
        logging.CRITICAL: ('yellow', 'on_red')
        },
    }
# _ENDOF_ configuration
"""Global execo configuration parameters.

- ``log_level``: the log level (see module `logging`)

- ``remote_tool``: default tool to use when instanciating remote
  processes. Can be `execo.config.SSH` or `execo.config.TAKTUK`

- ``fileput_tool``: default tool to use to put files remotely. Can be
  `execo.config.SCP` or `execo.config.TAKTUK`

- ``fileget_tool``: default tool to use to get remote files. Can be
  `execo.config.SCP` or `execo.config.TAKTUK`

- ``compact_output_threshold``: only beginning and end of stdout /
  stderr are displayed by `execo.process.ProcessBase.dump` when their
  size is greater than this threshold. 0 for no threshold

- ``kill_timeout``: number of seconds to wait after a clean SIGTERM
  kill before assuming that the process is not responsive and killing
  it with SIGKILL

- ``intr_period``: number of seconds between periodic check interrupt,
  for correct handling of ctrl-c

- ``port_range``: a tuple (start port, end port) of ports to use for
  the function ``execo.utils.get_port``. As all python ranges, start
  is inclusive, end is exclusive.

- ``kill_childs_at_end``: Whether to try sending SIGTERM to all
  subprocesses started through execo when the script
  terminates. Warnings:

  - this config option must be set at execo import time, changing it
    later will be ignored

  - SIGTERM is only sent to all childs or subchilds which did not try
    to daemonize by changing their process group.

- ``color_mode``: whether to colorize output (with ansi escape
  sequences)

- ``color_styles``: mapping of identifiers to iterables of ansi
  attributes identifiers (see `execo.log._ansi_styles`)
"""

def make_default_connection_params():
# _STARTOF_ default_connection_params
    default_connection_params = {
        'user':        None,
        'keyfile':     None,
        'port':        None,
        'ssh':         'ssh',
        'scp':         'scp',
        'taktuk':      'taktuk',
        'ssh_options': ( '-tt',
                         '-o', 'BatchMode=yes',
                         '-o', 'PasswordAuthentication=no',
                         '-o', 'StrictHostKeyChecking=no',
                         '-o', 'UserKnownHostsFile=/dev/null',
                         '-o', 'ConnectTimeout=20' ),
        'scp_options': ( '-o', 'BatchMode=yes',
                         '-o', 'PasswordAuthentication=no',
                         '-o', 'StrictHostKeyChecking=no',
                         '-o', 'UserKnownHostsFile=/dev/null',
                         '-o', 'ConnectTimeout=20',
                         '-rp' ),
        'taktuk_options': ( '-s', ),
        'taktuk_connector': 'ssh',
        'taktuk_connector_options': ( '-o', 'BatchMode=yes',
                                      '-o', 'PasswordAuthentication=no',
                                      '-o', 'StrictHostKeyChecking=no',
                                      '-o', 'UserKnownHostsFile=/dev/null',
                                      '-o', 'ConnectTimeout=20'),
        'chainput_nc_client_timeout': 5,
        'chainput_nc_server_timeout': 30,
        'nc': '/bin/nc.traditional -v -v',
        'chainput_host_retry': 5,
        'chainput_chain_retry': 10,
        'chainput_try_delay': 1,
        'forwarding_timeout': 25,
        'pty': False,
        'host_rewrite_func': None
        }
# _ENDOF_ default_connection_params
    return default_connection_params

default_connection_params = make_default_connection_params()
"""Default connection params for ``ssh``/``scp``/``taktuk`` connections.

- ``user``: the user to connect with.

- ``keyfile``: the keyfile to connect with.

- ``port``: the port to connect to.

- ``ssh``: the ssh or ssh-like command.

- ``scp``: the scp or scp-like command.

- ``taktuk``: the taktuk command.

- ``ssh_options``: tuple of options passed to ssh.

- ``scp_options``: tuple of options passed to scp.

- ``taktuk_options``: tuple of options passed to taktuk.

- ``taktuk_connector``: the ssh-like connector command for taktuk.

- ``taktuk_connector_options``: tuple of options passed to
  taktuk_connector.

- ``nc``: the netcat command to use

- ``chainput_nc_client_timeout``: timeout for client connection to
  next hop

- ``chainput_nc_server_timeout``: timeout for server to wait for
  incoming connection from previous hop

- ``chainput_host_retry``: number of times each hop in the transfer
  chain retries to connect to next hop

- ``chainput_chain_retry``: number of times each hop in the transfer
  chain tries a new next hop if all tries to current next hop fail. If
  given a float (between 0.0 and 1.0), this is expressed as a ratio of
  the total number of hosts in the chain.

- ``chainput_try_delay``: delay in seconds between TCP client to
  server connection attempts.

- ``forwarding_timeout``: max time to wait for a forwarding
  port to be available

- ``pty``: boolean. Wether to allocate or not a pty for
  ssh/scp.

- ``host_rewrite_func``: function called to rewrite hosts
  addresses. Takes a host address, returns a host address.
"""

def make_connection_params(connection_params = None, default_params = None):
    return_params = make_default_connection_params()
    if default_params == None:
        default_params = default_connection_params
    return_params.update(default_params)
    if connection_params: return_params.update(connection_params)
    return return_params

def load_configuration(filename, dicts_confs):
    """Update dicts with those found in file.

    :param filename: file to load dicts from

    :param dicts_confs: an iterable of couples (dict, string)

    Used to read configuration dicts. For each couple (dict, string),
    if a dict named string is defined in <file>, update
    dict with the content of this dict. Does nothing if unable to open
    <file>.
    """
    if os.path.isfile(filename):
        jailed_globals = {}
        try:
            with open(filename) as f:
                code = f.read()
                ccode = compile(code, filename, 'exec')
                exec(ccode, jailed_globals)
        except Exception as exc: #IGNORE:W0703
            print("ERROR while reading config file %s:" % (filename,))
            print(exc)
        for (dictio, conf) in dicts_confs:
            if conf in jailed_globals:
                dictio.update(jailed_globals[conf])

def get_user_config_filename():
    _user_conf_file = None
    if 'HOME' in os.environ:
        _user_conf_file = os.environ['HOME'] + '/.execo.conf.py'
    return _user_conf_file

__cf = get_user_config_filename()
if __cf:
    load_configuration(
        __cf,
        ((configuration, 'configuration'),
         (default_connection_params, 'default_connection_params')))
